Added support for reading/writing Madden 24 and Madden 25 rosters (table type 5 and field type 4)

WiiExpertise opened the pull request:
Added support for TDB2 table type 5 and field types 3 and 4, therefore adding support for Madden 24 and Madden 25 roster files. Also added a couple of test roster files accordingly. 

Additionally, fixed an issue with strings > 128 characters. The old implementation assumed the string length is always a single byte, but it is in fact a standard LEB number like other integers in the file, and it can span multiple bytes for longer strings.

Below is some info on what is different:

**Table type 5 (keyed record):**
- Used to store CharacterVisuals data
- Contains an extra byte after the table's unknown1 byte representing record data storage type 
- Each record is preceded by a LEB-compressed record index (this is used as the key value to match a POID from the player table to a record in this table). This table has an interesting quirk where the records need to be written in ascending record index order, otherwise it can cause the game to not read the records properly. My guess is that this is because they are using some kind of binary search implementation to look up records, so if they are not in order, the binary search doesn't work like it should.
- If data storage type is 0x03, record data is stored uncompressed
- If data storage type is 0x02, record data is stored with gzip compression
- For uncompressed records, format is essentially the same as a standard TDB2 record, with the exception that you may see field types 3 and 4, described below.
- For compressed records, between the compressed data and the record index, there is an LEB-compressed integer for the size in bytes of the compressed data. You then read that many bytes and use gunzip to get the TDB2 data. The decompressed data starts with a header key (CHVI) and unknown byte before proceeding into standard TDB record format. The last byte of the data is 0x00, just like a standard record.

**Field type 3 (unknown):**
- Unknown, used for the field "COMB". Usually has a value of 0, though M24 rosters are unique in that they don't seem to have a value at all for this field type

**Field type 4 (subtable):**
- Used to store array-like CharacterVisuals data (loadouts, etc.)
- Essentially a sub-table within a TDB2 record. Format is identical to that of a TDB2 table, with the 3 byte field key being the table key, followed by the table type byte, an unknown byte, and then the record count.

**Note:** It is unclear what results in this, but newer Madden 25 rosters, particularly official EA ones, have a weird quirk where not all possible fields are defined for each record. If the value of a field is 0 or empty for a record, they have started to not define the field at all for that record. Some roster files seem to be fine, while others have this quirk. Be careful with this.

bep713 commented on streams/TDB2/TDB2Parser.js:
```diff
@@ -25,19 +29,224 @@ class TDB2Parser extends FileParser {
+    };
+
+    _onCompressedTableFieldStart(compressedRecordBuf, record, table, existingParser) {
+        const decompressedRecordBuf = !existingParser ? zlib.gunzipSync(compressedRecordBuf) : null;
```
Would prefer using gunzip async here. Because the parser is wonky, you'd have to use the callback version and make a new function to handle the done case. Basically, it'd look like:

```
if (!existingParser) {
   zlib.gunzip(compressedRecordBuf, (err, data) => {
     this._onDecompressedTableFieldStart(data);
  }
} else {
   this._onDecompressedTableFieldStart(existingParser); // Not 100% sure this is right
}
```

and then everything below would be in that new function

WiiExpertise replied:
So unfortunately, it seems setting it up the way you describe causes `Error: got data but not currently parsing anything` because _onCompressedTableFieldStart calls the async gunzip and then returns, which causes the parser to not do anything while waiting for the gunzip to finish and then execute the callback.

bep713 replied:
Ah, that makes sense, I think I had to do that with another file type. I'm good with your changes, though I have some thoughts if you're interested in refactoring at some point:

That has me thinking that we shouldn't decompress in TDB2Parser. Is there an easy way to determine total compressed size and just read the entire compressed table all in one big chunk and store in some class?

The class (maybe TDB2CompressedTable or something, or even just TDB2Table) would have a function to decompress/parse itself with the same logic you wrote, which can be called after the TDB2Parser is done. It can be called either right after completion or on-demand if its quick enough (that way we spend time decompressing only if the user specifically requests the data). This is _kinda_ what madden-franchise does with `readRecords` and the strategies.

Would also probably make things easier because you don't have to worry about the odd Parser behavior.

WiiExpertise replied:
The length of the compressed data is simply a LEB number immediately before the compressed data, so it should be possible to just read the compressed data and store it somewhere. I like the idea of storing it in a compressed record data structure. Definitely something that can be looked into down the road.

bep713 commented on streams/TDB2/TDB2Parser.js:
```diff
@@ -25,19 +29,224 @@ class TDB2Parser extends FileParser {
+    _onCompressedTableFieldStart(compressedRecordBuf, record, table, existingParser) {
+        const decompressedRecordBuf = !existingParser ? zlib.gunzipSync(compressedRecordBuf) : null;
+
+        const recordParser = existingParser ? existingParser : new SimpleParser(decompressedRecordBuf);
```
Do you need the `SimpleParser` as opposed to Node's built in buffer methods? They aren't as nice to use as C#'s but they should be able to do the same thing. Would simplify this a bit I think.

WiiExpertise replied:
While I imagine it probably could be done using buffer methods, I'm not sure that it would actually be simpler to do so. I opted to use the SimpleParser module I put together in this case as it was easier than maintaining a buffer and managing the current read offset manually, especially given the potential recursive case for reading subtables.

bep713 commented on streams/TDB2/TDB2Parser.js:
```diff
@@ -25,19 +29,224 @@ class TDB2Parser extends FileParser {
+            case FIELD_TYPE_INT:
+                field.raw = utilService.writeModifiedLebCompressedInteger(utilService.parseModifiedLebEncodedNumber(recordParser));
+                record.fields[field.key] = field;
+                if(field.key === 'UNWI')
```
Fun! Any other way to detect this case besides the key name?

WiiExpertise replied:
I can't really think of anything concrete. I thought of potentially checking if the next byte's leading bit is set, since this should always be the case if the next byte is part of the next 6bit compressed field key, but this would fall apart if the UNWI key happened to come at the end of a record, since the next byte would be 0 in that case.

But I did determine that this field seems to only exist in M24 rosters, not M25 rosters, so I removed it from the compressed record field reading in the latest commit.

bep713 commented on streams/TDB2/TDB2Parser.js:
```diff
@@ -25,19 +29,224 @@ class TDB2Parser extends FileParser {
+                }
+            }
+
+            table.records.push(new Proxy(record, {
+                get: function (target, prop, receiver) {
+                    return record.fields[prop] !== undefined ? record.fields[prop].value : record[prop] !== undefined ? record[prop] : null;
+                },
+                set: function (target, prop, receiver) {
+                    if (record.fields[prop] !== undefined) {
+                        record.fields[prop].value = receiver;
+                    }
+                    else {
+                        record[prop] = receiver;
+                    }
+
+                    return true;
+                }
+            }));
```
Awesome that you used a Proxy, could potentially extract this into a module to avoid duplicating the code below

WiiExpertise replied:
Done in latest commit, using a new _pushTableRecord function

bep713 commented:
Thanks for adding documentation & test data. Could you add some test cases for the subtable/compressed data for when M26 inevitably changes everything? Lol

WiiExpertise commented:
Cleaned up some things and also addressed the reviewed items that I could. All un-addressed reviewed items were justified in a reply.

Added a testing case for table type 5 in both M24 and M25 rosters in tests/tdb2Tests.js. It's nothing too crazy, but for both M24 and M25, it loads the test roster and sets Geno Smith's (first record in the CVPM and PLEX tables respectively) right arm sleeve to all white. This goes a couple subtables deep, so it should be a good test of the table type 5 reading, both compressed and uncompressed, and the subtable reading.

bep713 commented on tests/tdb2Tests.js:
```diff
@@ -0,0 +1,91 @@
+const m24Path = 'data/M24_ROSTER-Official';
```
Looks good. Can you move this file into the `e2e` folder?

WiiExpertise replied:
Done.

bep713 commented on package.json:
```diff
@@ -31,6 +31,7 @@
     "leb": "^0.3.0",
     "lz4": "^0.6.5",
     "memoryjs": "^3.4.0",
+    "stream": "^0.0.3",
```
Is this needed?

WiiExpertise replied:
Not sure why that's there, I didn't explicitly add it.

WiiExpertise commented:
Moved TDBTests file to e2e folder, and updated save locations accordingly.

Also resolved the issue with non-existent fields in some M25 TDB2 files by adding all fields to every record. In the process of doing this, overhauled the logic for table fieldDefinitions. Instead of just looking at the fields of the first record to determine fieldDefinitions, they are now determined during record parsing. When reading each field in a record, it will check if fieldDefinitions has that field key already. If it doesn't, it will add it. This ensures that every possible field in a table has an entry, even if the first record didn't have that field.

After completing reading of a table, it goes back through every record and adds any missing fields, to make sure every record has every field. This is what resolves the non-existent field issue.

bep713 commented on tests/e2e/TDB2Tests.js:
Instead of writing a script for testing, it's better to use a testing framework like mocha & chai to assert things. That way, you can run all of tests at once pre-commit to make sure that other functionality didn't accidentally break. Take a look at `TDBFile.spec.js` for a pretty simple test file using mocha & chai.

You can run all of the tests with `mocha tests/**/*.spec.js` or you can run a specific test by specifying the path instead of the wildcards.

bep713 commented on tests/e2e/TDB2Tests.js:
```diff
@@ -0,0 +1,91 @@
+    const m25Helper = new MaddenRosterHelper();
+    
+    // These two tests change Geno Smith's right arm sleeve to all white in Madden 24 and Madden 25 respectively
+    m24Helper.load(m24Path)
```
Example: 
```
    it('M24 Helper test', () => {
       m24Helper.load(m24Path).then(() => {
          // ...paste everything here
       }
    });
```

bep713 commented on tests/e2e/TDB2Tests.js:
```diff
@@ -0,0 +1,91 @@
+
+            m24Helper.save("../data/WriteTest_M24_ROSTER-Official")
+                .then(() => {
+                    console.log("Completed M24 CVPM test. Saved file to ../data/WriteTest_M24_ROSTER-Official");
```
assert that the values are appearing as expected here w/ chai so that you don't need to manually verify. With save tests, that usually means re-loading the file and verifying all data is expected.

Here's an example that you'll need to modify a bit but hopefully demonstrates what i mean
```
const newHelper = new MaddenRosterHelper();  // may not be necessary
newHelper.load("../data/WriteTest_M24_ROSTER-Official").then(() => {
   const newGeno = m24Helper.file.CVPM.records.find((record) => {
         return record.fields['ASNM'].value === "SmithGeno_112";
   });

   expect(newGeno.fields['LOUT'].value).to.equal(something);
   // or if you want to compare to an object
   expect(newGeno.fields['LOUT'].value).to.eql({ something: 'something' });
});
```

WiiExpertise commented:
Added a simple test using chai for reading/writing to the table type 5.

WiiExpertise commented:
Discovered that M25 rosters have decided to be unique in their own way (what else is new). I previously thought field type 3 didn't have a value after it only in M24 rosters, but it appears some (but importantly, not all) M25 rosters have this trait as well. So I added a simple check to see if the byte after the field type is 0 and reading it if so. In practice, field type 3 never seems to appear at the end of a record, so this should be fine, at least for now.

Also, I swear the mocha test was working when I first pushed it, but when I went to use it now, it didn't like the path to the test data that I had. So I updated that to use the correct relative path and it works now.

WiiExpertise commented:
- Updated the TDB2 writer to sort table type 5 records by record index (used as the key) before writing (based on experience this is necessary for the game to read the records properly, as mentioned above I'm guessing this is because of it using binary search or something similar)
- Fixed an issue where the compressed record storage version of table type 5 was not having its field definitions populated in both it and any subtables

Added records:
- Added a new `deepCopyRecord` function to TDB2Record that allows for creating a deep copy of an existing record. **It is up to the developer to determine if there are any fields/properties in the record that must be unique for each record in the table and update them in the copy accordingly.**
- Added a new `addRecord` function to TDB2Table, which allows for adding a provided TDB2Record object to a table, allowing for adding new records to TDB2 tables. 
- Included some test cases for adding new records for both M24 and M25 files.

bep713 commented on filetypes/TDB2/TDB2Record.js:
```diff
@@ -23,6 +23,63 @@ class TDB2Record {
+        }
+
+        // Handle circular references
+        if (cache.has(record)) {
+            return cache.get(record);
+        }
```
nice use of memoization!

bep713 commented on filetypes/TDB2/TDB2Table.js:
```diff
@@ -1,4 +1,6 @@
 const utilService = require('../../services/utilService');
+const TDB2Record = require('./TDB2Record');
+const lodash = require('lodash');
```
```suggestion
```

bep713 commented on streams/TDB2/TDB2Parser.js:
```diff
@@ -151,6 +165,18 @@ class TDB2Parser extends FileParser {
                 field.key = utilService.getUncompressedTextFromSixBitCompression(field.rawKey.slice(0, 3));
                 field.type = field.rawKey.slice(3).readUInt8(0);
 
+                if (!table.fieldDefinitions.find((f) => f.name === field.key)) {
+                    const newFieldDef = {
+                        'name': field.key,
+                        'type': field.type,
+                        'offset': -1,
+                        'bits': -1,
+                        'maxValue': -1
+                    }
+    
+                    table.fieldDefinitions.push(newFieldDef);
+                }
```
I'd make this and lines 96-106 into a shared function to avoid code duplication

WiiExpertise replied:
Done in latest commit

bep713 reviewed (changes requested):
Looks good! Just a few small tweaks needed

WiiExpertise commented:
Removed the lodash requirement that got left in from testing, oops.

Also moved the field definition population to a separate function in TDB2Parser.

bep713 commented on package.json:
```diff
@@ -54,5 +54,5 @@
     "url": "https://github.com/bep713/madden-file-tools.git"
   },
-  "version": "2.7.4"
+  "version": "2.7.5"
```
```suggestion
  "version": "2.8.0"
```

bep713 commented on package.json:
```diff
@@ -31,6 +31,7 @@
     "leb": "^0.3.0",
     "lz4": "^0.6.5",
     "memoryjs": "^3.4.0",
+    "stream": "^0.0.3",
```
```suggestion
```
